capytaine/bem/engines.py

```diff
@@ -315,3 +327,144 @@ def get_col_func(j):
                 mesh1, mesh2, free_surface, water_depth, wavenumber,
             )
             return S, V
+
+class HierarchicalPrecondMatrixEngine(HierarchicalToeplitzMatrixEngine):
+    """An experimental matrix engine that build a hierarchical matrix with
+     some block-Toeplitz structure.
+
+    Parameters
+    ----------
+    ACA_distance: float, optional
+        Above this distance, the ACA is used to approximate the matrix with a low-rank block.
+    ACA_tol: float, optional
+        The tolerance of the ACA when building a low-rank matrix.
+    matrix_cache_size: int, optional
+        number of matrices to keep in cache
+    """
+
+    def __init__(self, *, ACA_distance=8.0, ACA_tol=1e-2, matrix_cache_size=1):
+
+        if matrix_cache_size > 0:
+            self.build_matrices = delete_first_lru_cache(maxsize=matrix_cache_size)(self.build_matrices)
+
+        self.ACA_distance = ACA_distance
+        self.ACA_tol = ACA_tol
+
+        self.linear_solver = linear_solvers.solve_precond_gmres
+
+        self.exportable_settings = {
+            'engine': 'HierarchicalToeplitzMatrixEngine',
+            'ACA_distance': ACA_distance,
+            'ACA_tol': ACA_tol,
+            'matrix_cache_size': matrix_cache_size,
+        }
+
+    def access_block_by_path(self, matrix, path):
```

Comment:
Since this does not depends on the `MatrixEngine` (`self` is unused), this could just be a method of `BlockMatrix`.